container_crawler/__init__.py

```
@@ -86,40 +139,73 @@ def _worker(self):
                 continue
 
             try:
-                if work:
-                    row, handler = work
-                    with self._swift_pool.item() as swift_client:
-                        handler.handle(row, swift_client)
+                if not work:
+                    break
+
+                row, handler = work[0]
+                container_job = work[1]
+                with self._swift_pool.item() as swift_client:
+                    handler.handle(row, swift_client)
+                container_job.complete_task()
             except RetryError:
-                self.error_queue.put((row, None))
+                container_job.complete_task(retry=True)
             except:
-                self.error_queue.put((row, traceback.format_exc()))
+                container_job.complete_task(error=True)
+                self.log('error', u'Failed to handle row %s (%s): %r' % (
+                    row['ROWID'], row['name'].decode('utf-8'),
+                    traceback.format_exc()))
             finally:
                 self.work_queue.task_done()
 
-    def _stop(self):
-        for _ in xrange(self.workers):
-            self.work_queue.put(None)
-        self.pool.waitall()
-
-    def _check_errors(self):
-        # When working in bulk, errors are propagated immediately
-        if self.bulk or self.error_queue.empty():
-            return
+    def _enumerator(self):
+        job = ContainerJob()
+        while 1:
+            try:
+                work = self.enumerator_queue.get()
+            except:
```

should this at least be `except Exception:` or does it really need to catch-and-ignore exceptions not inheriting from `Exception`?

---

At some point I remember being worried about an exception being thrown that does not inherit from `Exception`. I think specifically I was concerned about `GeneratorExit`, although looking back I can't find documentation for this. It does align with the `_worker` method for now. I'd defer changing it to a separate change -- I think you're right that we probably should catch `Exception` here.